packages/flutter/lib/src/rendering/sliver_multi_box_adaptor.dart

```diff
@@ -224,23 +241,66 @@ abstract class RenderSliverMultiBoxAdaptor extends RenderSliver
 
   bool _debugAssertChildListLocked() => childManager.debugAssertChildListLocked();
 
+  /// Verify the child list index is in strictly increased order.
+  ///
+  /// This is a no-op for non-debug builds.
+  void _debugVerifyChildOrder(){
+    assert(() {
+      if (_debugChildOrderVerificationEnabled) {
+        int index = indexOf(firstChild);
+        RenderBox child = childAfter(firstChild);
+        while (child != null) {
+          assert(indexOf(child) > index);
+          index = indexOf(child);
+          child = childAfter(child);
+        }
+      }
+      return true;
+    }());
+  }
+
   @override
   void insert(RenderBox child, { RenderBox after }) {
     assert(!_keepAliveBucket.containsValue(child));
     super.insert(child, after: after);
     assert(firstChild != null);
     assert(() {
-      int index = indexOf(firstChild);
-      RenderBox child = childAfter(firstChild);
-      while (child != null) {
-        assert(indexOf(child) > index);
-        index = indexOf(child);
-        child = childAfter(child);
-      }
+      _debugVerifyChildOrder();
       return true;
     }());
   }
 
+  @override
+  void move(RenderBox child, { RenderBox after }) {
+    // There are two scenarios:
+    //
+    // 1, The child is not keptAlive.
+    // The child is in childList maintained by ContainerRenderObjectMixin.
+    // We can call the super.move and update parentData with new slot.
+    //
+    // 2, The child is keptAlive.
+    // In this case, the child is no longer in the childList but might be stored in
+    // [_keepAliveBucket]. We need to update the location of the child in the bucket.
+    final SliverMultiBoxAdaptorParentData childParentData = child.parentData;
+    if (!childParentData.keptAlive) {
+      super.move(child, after: after);
+      // The following call updates the slot in the parentData.
+      childManager.didAdoptChild(child);
+    } else {
+      // If the child in the bucket is not current child, that means someone has
+      // already replaced us. We cannot remove it.
+      if (_keepAliveBucket[childParentData.index] == child) {
+        _keepAliveBucket.remove(childParentData.index);
+      }
+      childManager.didAdoptChild(child);
+      // If there is an existing child in the new slot, that mean that child will
+      // be moved to other index. In other cases, the existing child should have been
+      // removed by updateChild. Thus, it is ok to overwrite it.
```

Comment:
We could have a `List _debugDangelingKeepAlives` to which we add (in an assert) children from the keepAliveBucket that are overwritten. When they get re-inserted in the keepAliveBucket we remove them from the list. When we turn the integrity checks back on we check that the list is empty.

Reply:
done